File: 알고리즘 스터디/개인공부/BFS&DFS/ABCDE.py
# https://www.acmicpc.net/problem/13023
import sys
sys.setrecursionlimit(20001)


def dig(people, visited, cur_node, remain):
    # 깊이가 4 이상이 되면 종료
    if remain == 0:
        print(1)
        exit(0)

    visited[cur_node] = True
    for neighbor in people[cur_node]:
        if visited[neighbor] is False:
            dig(people, visited, neighbor, remain - 1)

    visited[cur_node] = False


if __name__ == '__main__':
    n_people, n_line = map(int, sys.stdin.readline().split())
    people = [list() for _ in range(n_people)]
    answer = 0

    for _ in range(n_line):
        src, dst = map(int, sys.stdin.readline().split())
        people[src].append(dst)
        people[dst].append(src)

    for node in range(n_people):
        dig(people, [False] * n_people, node, 4)

    print(0)

File: 알고리즘 스터디/개인공부/BFS&DFS/ABCDE.cpp
#include<iostream>
#include<vector>
#include<cstring>

using namespace std;


void dig(vector<int> people[], int n_people, bool* visited, int cur_node, int remain) {
	vector<int>::iterator it;

	if (remain == 0) {
		cout << 1 << endl;
		exit(0);
	}

	visited[cur_node] = true;
	for (it = people[cur_node].begin(); it != people[cur_node].end(); it++) {
		if (!visited[*it]) {
			dig(people, n_people, visited, *it, remain - 1);
		}
	}
	visited[cur_node] = false;
}


int main() {
	int n_people, n_line, answer = 0;
	int src, dst;
	vector<int>* people;
	bool *visited;

	cin >> n_people >> n_line;
	people = new vector<int>[n_people];
	visited = new bool[n_people];

	for (int i = 0; i < n_line; i++) {
		cin >> src >> dst;
		people[src].push_back(dst);
		people[dst].push_back(src);
	}

	for (int i = 0; i < n_people; i++) {
		memset(visited, false, n_people * sizeof(bool));
		dig(people, n_people, visited, i, 4);
	}

	cout << answer << endl;

	return 0;
}